### ISSUE TITLE
Delay JS - Links needs to be clicked twice on iOS / Safari

### ISSUE THREAD

**Issue by arunbasillal** [wp-media/wp-rocket#3142]
**Before submitting an issue please check that you’ve completed the following steps:**
- Made sure you’re on the latest version 👍 
- Used the search feature to ensure that the bug hasn’t been reported before 👍 

**Describe the bug**
On Safari browser in iOS (and probably in others as per one customer), when delay JS is enabled links have to be clicked twice for them to work. This is even when:
- there aren't any keywords in the delay JS text area.
- link preloading is disabled.
- tested on latest version of iOS and Safari. 

**To Reproduce**
- Enable delay JS
- In an iPhone or in BrowserStack use Safari to visit the page.
- Scroll down (important) and then try to click on a link. Only the second click will register and open the page.

Note that if you click on a link which is in the viewport, i.e. without scrolling, it will open immediately.

**Expected behavior**
Links should load as normal and shouldn't need two clicks. 

**Screenshots**
Screencast: https://youtu.be/D1Pp45wLMhE (Thanks to @vmanthos for this and most of this report)

**Additional context**
Please see this note from Vasilis for some more insights - https://secure.helpscout.net/conversation/1283702266/195256/#thread-3691038967

Potentially related tickets: 
- https://secure.helpscout.net/conversation/1287155173/196167/
- https://secure.helpscout.net/conversation/1283702266/195256/ (Same ticket linked above)
- https://secure.helpscout.net/conversation/1277735438/193444/
- https://secure.helpscout.net/conversation/1277423432/193297/

**Backlog Grooming (for WP Media dev team use only)**
- [ ] Reproduce the problem
- [ ] Identify the root cause
- [ ] Scope a solution
- [ ] Estimate the effort

**Comment by socialpreneur** [comment]
I had a client reporting this with Flying Scripts plugin, too.

**Comment by engahmeds3ed** [comment]
After searching I got that there is an issue on IOS with Safari related to using `touchstart` or `touchmove` without using `touchend` which make the links double click issue ( you need to click on any point from the screen as first click then u can click on any link )

the following link tries to highlight the issue:
https://stackoverflow.com/a/10340968

On customer site I made the following change and solved the issue:
https://github.com/wp-media/wp-rocket/blob/7ec9d996f004707a57c4c8476090d5bd943051f0/assets/js/lazyload-scripts.js#L91-L99
replace `touchmove` with `touchmove touchend`
replace `touchstart` with `touchstart touchend`

After checking with @hellofromtonya , she said:-

> This particular issue will need more investigation to find out why. Rather than adding a patch to the code for this specific edge case, why do they have the issue and why does this specific combination fix it?
I think the “why” here will be very valuable to ensure we don’t inject another problem for other customers.

so I will add needs: r&d label to be investigated.

**Comment by webtrainingwheels** [comment]
Another case: https://secure.helpscout.net/conversation/1296370691/198532?folderId=377611
Customer reports it happening on chome and safari on mobile

**Comment by vmanthos** [comment]
**Related ticket:** https://secure.helpscout.net/conversation/1296809928/198688/

**Comment by vmanthos** [comment]
**Related ticket:** https://secure.helpscout.net/conversation/1322836283/206594?folderId=2135277

**Comment by girlie** [comment]
**Related ticket:** https://secure.helpscout.net/conversation/1334490646/210819/

**Comment by NataliaDrause** [comment]
**Related ticket:** https://secure.helpscout.net/conversation/1336672544/211528/
Notes and screen recording in the notes. The issue happens in both Safari and Chrome browsers.

**Comment by Roboonl** [comment]
> After searching I got that there is an issue on IOS with Safari related to using `touchstart` or `touchmove` without using `touchend` which make the links double click issue ( you need to click on any point from the screen as first click then u can click on any link )
> 
> the following link tries to highlight the issue:
> https://stackoverflow.com/a/10340968
> 
> On customer site I made the following change and solved the issue:
> https://github.com/wp-media/wp-rocket/blob/7ec9d996f004707a57c4c8476090d5bd943051f0/assets/js/lazyload-scripts.js#L91-L99
> 
> 
> replace `touchmove` with `touchmove touchend`
> replace `touchstart` with `touchstart touchend`
> After checking with @hellofromtonya , she said:-
> 
> > This particular issue will need more investigation to find out why. Rather than adding a patch to the code for this specific edge case, why do they have the issue and why does this specific combination fix it?
> > I think the “why” here will be very valuable to ensure we don’t inject another problem for other customers.
> 
> so I will add needs: r&d label to be investigated.

I tried this suggestion but it unfortunately didn't resolve the issue on iOS (for me at least).

**Comment by BradFD** [comment]
Same problem. Tried this feature for the first time a few days ago and got immediate reports that links on iOS weren't working and they needed double tap to work. Tried on 2 x iOS 14 devices and confirmed this problem so disabled for now.

**Comment by Roboonl** [comment]
Any updates on this?

**Comment by NataliaDrause** [comment]
Related: https://secure.helpscout.net/conversation/1367506767/223234

**Comment by veganWebagency** [comment]
Same problem. Any updates?

**Comment by kashishkumawat** [comment]
Experiencing the same bug with both WP-Rocket and Flying Scripts. Replacing `touchstart` and `touchmove` didn't help.

**Comment by jorgemartine00** [comment]
Related: https://secure.helpscout.net/conversation/1376564315/226026/

**Comment by socialpreneur** [comment]
I have this happening on iOS Chrome browser rather than Safari. Anyone else can confirm?

**Comment by Roboonl** [comment]
We've dropped the WP Rocket plugin for now, this issue is already open for 3 months and I don't feel like it's going to be fixed any time soon with the current open issue count (+400).

**Comment by NataliaDrause** [comment]
Related: https://secure.helpscout.net/conversation/1379195337/226677

**Comment by roberthedlund** [comment]
Hope WP-rocket add the "fix" soon :)

**Comment by veganwebagency** [comment]
Any update from the developers? Are you working on it?

**Comment by arunbasillal** [comment]
@wp-media/dev Since we are seeing more and more of this and since the bug makes the feature quite unusable, I think we need to prioritise it. I have added it to the next sprint to see if we can figure out a solution. 🙏

**Comment by roberthedlund** [comment]
> @wp-media/dev Since we are seeing more and more of this and since the bug makes the feature quite unusable, I think we need to prioritise it. I have added it to the next sprint to see if we can figure out a solution. 🙏

Sounds good. This feature make a big difference for my website when it comes to speed, would love to have it activated again =)

**Comment by Tabrisrp** [comment]
@arunbasillal no, the issue is marked as need R&D, it can't be part of a sprint. For an issue to be part of a sprint, it needs to be fully groomed first.

**Comment by veganwebagency** [comment]
> @wp-media/dev Since we are seeing more and more of this and since the bug makes the feature quite unusable, I think we need to prioritise it. I have added it to the next sprint to see if we can figure out a solution. 🙏

Thanks! I really love this feature - would be awesome if it can be used on all my websites!

**Comment by Tabrisrp** [comment]
@wp-media/productrocket Some feedbacks from the team about this:
- This is tricky because of the way iOS works.
- The thing is touchscreens don’t really have a good hover alternative, other than maybe touch-and-hold — but at least on iOS devices, that has its own significance (it loads a preview frame of the link destination).
- Ahmed deeply dived into this also before, the only solution is to use timeout like flying scripts

As there is no fix for that without significant changes to the current way this feature works, it needs further thinking before going further.

**Comment by Garnet-Fox** [comment]
Need and mouseWheel event and setTimeOut. On mouseweel scripts like Infinite Scroll and others scripts, works not correct, not loading next page if only mousewhell. In general, this applies to all scripts, and it critical.
Delay JS without setTimeOut also cause problems, in addition to links needs to be clicked twice. Exemple: In the Russian segment, Yandex search engine. The counter can use a webviewer, this is a record of user actions. Accordingly, the script should not be delayed: since when viewing user actions, the counter script on the site should be loaded. And it naturally does not load because no actions, no 'keydown', 'mouseover', 'touchmove', 'touchstart' occur in the viewer.

**Comment by socialpreneur** [comment]
I'd really like to know how WP Rocket team is considering this, but at least auto load scripts after x seconds option is a must until iOS fixes the issue.

**Comment by veganwebagency** [comment]
> @wp-media/productrocket Some feedbacks from the team about this:
> 
> * This is tricky because of the way iOS works.
> * The thing is touchscreens don’t really have a good hover alternative, other than maybe touch-and-hold — but at least on iOS devices, that has its own significance (it loads a preview frame of the link destination).
> * Ahmed deeply dived into this also before, the only solution is to use timeout like flying scripts
> 
> As there is no fix for that without significant changes to the current way this feature works, it needs further thinking before going further.

Any timeline for the implementation for the timeout solution?

**Comment by GeekPress** [comment]
Loading scripts after X seconds isn't the solution. Doing that is just in the opposite of the interest of this feature. If we can find a solution without doing that, we will try to go on this way.

@arunbasillal @engahmeds3ed Could you redo some tests?

I'm asking as we had the same problem with our website too. But since last week, we don't have the problem anymore. Is it the coincidence of releasing our new website, or an iOS update?

> Any timeline for the implementation for the timeout solution?

We can't provide any ETA.

**Comment by socialpreneur** [comment]
@GeekPress When I view your site with Chrome using mobile window size provided by Dev tool (Is there any official name for this feature? Perhaps mobile preview?), two finger scrolling never triggers delayed scripts including analytics. And for the double click problems on this thread, which page and link did you test? The top page?

**Comment by robin-onlineidentity** [comment]
> @GeekPress When I view your site with Chrome using mobile window size provided by Dev tool (Is there any official name for this feature? Perhaps mobile preview?), two finger scrolling never triggers delayed scripts including analytics. And for the double click problems on this thread, which page and link did you test? The top page?

I don't think testing with the Mobile Preview in the Chrome Developer Tool gives the same result as actually testing on an iOS device due to the difference in touch gestures.

**Comment by Roboonl** [comment]
> @GeekPress When I view your site with Chrome using mobile window size provided by Dev tool (Is there any official name for this feature? Perhaps mobile preview?), two finger scrolling never triggers delayed scripts including analytics. And for the double click problems on this thread, which page and link did you test? The top page?

I don't think testing with the Mobile Preview in the Chrome Developer Tool gives the same result as actually testing on an iOS device due to the difference in touch gestures.

**Comment by socialpreneur** [comment]
@Roboonl You are right, they should be different. And I'm not doing so. I simply wonder if it is OK for WP Rocket that scripts are never loaded under certain condition(s).

**Comment by Garnet-Fox** [comment]
The setTimeOut is needed, or rather necessary.

**Comment by NataliaDrause** [comment]
related: https://secure.helpscout.net/conversation/1406743456/234374/

**Comment by sandyfzu** [comment]
Related case: https://secure.helpscout.net/conversation/1405169718/233859/

Browser: Safari and chrome
Device: iPhone 11 Pro

**Comment by piotrbak** [comment]
Adding timeout trigger would also resolve https://github.com/wp-media/wp-rocket/issues/3420

**Comment by GeekPress** [comment]
> Adding timeout trigger would also resolve #3420

@piotrbak Not sure #3420 is related to this problem. Fixing #3454 should I think fix #3420

**Comment by piotrbak** [comment]
@GeekPress From my understanding, the users were scrolling around on the webpage and the Delay JS wasn't triggered. The Google Analytics script was not loaded then. 

Seems that you're right about the https://github.com/wp-media/wp-rocket/issues/3454 it's also related to mobiles. Probably any of those issues would resolve the https://github.com/wp-media/wp-rocket/issues/3420 
Thanks

**Comment by roberthedlund** [comment]
Any update?

**Comment by roberthedlund** [comment]
@arunbasillal  Cant you add an option in WP-rocket for this to exclude on certain pages instead in the meanwhile?
Like Flying Scripts has, they even have option to exclude the Delay JS for certain pages with keywords.

**Comment by alfonso100** [comment]
another case: 
https://secure.helpscout.net/conversation/1417928403/237257?folderId=2683093

**Comment by GeekPress** [comment]
> I don't think testing with the Mobile Preview in the Chrome Developer Tool gives the same result as actually testing on an iOS device due to the difference in touch gestures.

@Roboonl It has been tested on a real mobile device. Please try our website on mobile, you don't need to click two times.

@roberthedlund I should have missed something, I don't find any way with Flying Script to have keywords based on pages.

**Comment by roberthedlund** [comment]
> > I don't think testing with the Mobile Preview in the Chrome Developer Tool gives the same result as actually testing on an iOS device due to the difference in touch gestures.
> 
> @Roboonl It has been tested on a real mobile device. Please try our website on mobile, you don't need to click two times.
> 
> @roberthedlund I should have missed something, I don't find any way with Flying Script to have keywords based on pages.

[https://ibb.co/qYVyhh6](url)

**Comment by GeekPress** [comment]
@roberthedlund Well, you can already do that with WP Rocket too. You have just to go on your page edit screen, open the « WP Rocket Options » metabox, and turn off the Delay JS Execution option.

**Comment by roberthedlund** [comment]
> @roberthedlund Well, you can already do that with WP Rocket too. You have just to go on your page edit screen, open the « WP Rocket Options » metabox, and turn off the Delay JS Execution option.

Yeah but not ideal to do that manually on 100+ pages for an example.

**Comment by veganwebagency** [comment]
Any update?

**Comment by socialpreneur** [comment]
This issue is still available with version 3.8.6 with Chrome on Android. I just tried Fying Scripts, then this double click bug didn't happen. For iOS, even with Flying Scripts, viewers are required to double click.

**Comment by roberthedlund** [comment]
Any update? I find it strange that you have still not informed the users of WP-rocket in any way about the problem. How difficult is it to enter a warning notice before people activate the Delay JS feature? Have seen so many users in various forums who do not yet know that they are suffering from the problem before I have contacted them about it. It's a pretty big problem and I think you should warn about it.

**Comment by GeekPress** [comment]
@socialpreneur For iOS, are you trying on a real mobile, and not with a simulation?

Why I'm asking the question?

I retry again myself on my personal iPhone 8 and I don't have to click twice on all websites I tried which had Delay JS including our own website.

**Comment by socialpreneur** [comment]
@GeekPress On real mobile, including me and my client's phone, so I had to disable Delay JS completely. Flying Scripts didn't also work, so I'm guessing this is iOS limitation right now.

Added: It looks like there's some external or internal script which is causing this problem. Not every site which uses Delay JS of WP Rocket nor Flying Scripts have this double click symptoms. So far two sites that do not have this problem do not use adsense. I think we'll need to gather more cases to figure out which is the cause.

**Comment by roberthedlund** [comment]
@GeekPress 
> @socialpreneur For iOS, are you trying on a real mobile, and not with a simulation?
> 
> Why I'm asking the question?
> 
> I retry again myself on my personal iPhone 8 and I don't have to click twice on all websites I tried which had Delay JS including our own website.

I had tried on plenty of websites that using Delay JS, and have tried on real phones. iPhone 8, iPhone SE, iPhone 11 and all of them had this problem. Sometimes it workes on first "click" then the others totally stop working. So the issue is real.

**Comment by GeekPress** [comment]
@socialpreneur Have you tried on our website (https://wp-rocket.me)? As I said, I'm not able to reproduce the issue after trying on a couple of websites.

May I know the iOS version you are using?

**Comment by socialpreneur** [comment]
@GeekPress Not everything you see is the truth. If at least one WP Rocket user is having the issue, then we need to do everything and anything to solve the matter. My iOS is 14.4, and I'll ask my client for his iOS version. As I said, since this is happening on every Delay JS plugins out there (WP Rocket, Flying Scripts, Perfmatters) on certain site with certain scripts, this is not only your cause.

It may be browser, but when I see this double click problem, it's not only iOS, but I have it on my Android and its Chrome so it's clearly not only the iOS nor Safari problem. I just hope we can come up with some little hack or additional functions.php to bypass this trouble on certain sites.

I'll let you know if I find anything.

Added 1: OK, here's third site which is having the problem. Maybe I can narrow down some common scripts.

Added 2: So far just enabling Delay JS won't trigger this problem.
![delayjs](https://user-images.githubusercontent.com/1147374/111355140-6b800580-86ca-11eb-9c66-437ee52bd81d.png)

But when I set some scripts, double click/tap starts happening.
![delayjs2](https://user-images.githubusercontent.com/1147374/111355386-ac781a00-86ca-11eb-8c3b-c3ce9c3de5ff.png)

Added 3: Alright, here's what I found. When I add "table-of-contents" double click thing gets triggered. It may not be the same for all other users who are reporting this, but so far this is one of the keywords. Table of Contents Plus is the plugin name.

Added 4: Here's good news. Another site having the same problem is also using TOC Plus. @GeekPress, please test it with this plugin enabled. And tell all your support staffs to confirm if the users having the same problem is also using this plugin on their sites. ASAP.

Added 5: I have managed to solve this problem for one of the sites by not delaying TOC Plus script. But the same solution didn't work for another site, so there seems some more scripts that are not delay friendly.

Added 6: I have a feeling jquery version may be taking some part of this problem. One working site uses the latest, and the other not working site uses old version (/wp-includes/js/jquery/jquery.js?ver=1.12.4-wp). Debugging further more...

**Comment by socialpreneur** [comment]
@GeekPress Aside from TOC Plus, I also found that adsense is the cause of double click/tap. I have tested with a few sites, and all sites with different themes/plugins have the same. This happens with any adsense keywords below.

- googlesyndication
- adsbygoogle
- adsbygoogle.js

To reproduce

1. Build a test site with adsense anywhere, within above the fold or not.
2. Refresh the page, tap on header logo/text or mobile menu icon. At this point adsense gets loaded, but it won't open the menu with one tap.

Solution

I'm not sure as this seems how adsense script is designed. Considerable work around are:

1. Load adsense always after 1 sec. so most users won't face this problem.
2. Preload adsense on mouse hover (not sure if this would work on mobile tap)
3. Tell users not to monetize with adsense, so sites gets much faster...

**Comment by paskalr1** [comment]
@GeekPress my website is celmetro.com and I have the same double click problem when I enable Delay Javascript Execution, does not matter if there are no scripts to delay in the list.

**Comment by NataliaDrause** [comment]
Related: https://secure.helpscout.net/conversation/1458853194/249184?folderId=3864740
They use TOC Plus and adsense.
Edited: this is additional info from @socialpreneur with screen recording on the affected site.

**Comment by NicoCaldo** [comment]
I can confirm the issue on my website via Chrome mobile (as well as via Chrome Desktop setting the web dev tool to load as mobile)

**Comment by mikobig** [comment]
Same problem. Need to click twice on links. Happens on iOS devices with Safari, Chrome, Mozilla and Edge browsers. I have tested a lot and found a handful of scripts which do not trigger the problem when delayed.

**Comment by socialpreneur** [comment]
platform.twitter.com/widgets.js seems to be triggering double tap on Chrome with iOS. But there're sites that work fine, so I'm thinking depending on some themes, delaying js triggers double tap.

**Comment by socialpreneur** [comment]
Anybody know if iOS is planning to solve this issue? WP Rocket, Flying Scripts, Perfmatter's Delay JS, and all of these are having this issue.

**Comment by NickNoordijk** [comment]
Hey, found this through Google but I don't use wp-media or wp-rocket. The issue for me happens when I load the JS scripts of my advertiser after the first scroll. I need to tap the link twice to navigate. One tap anywhere on the screen followed by the link tap works too. When I load them at init the links can be tapped normally just once. Just wanted to let you all know it's not only these scripts. The scripts have a small issue but it seems iOS is the big issue here? 

UPDATE: I was able to fix it with @engahmeds3ed proposal! I added `touchend` to all my `touchstart` events and it fixed the bug! I've been looking for this for over a week!

**Comment by socialpreneur** [comment]
@NickNoordijk Did you use custom functions, or did you manually fix the plugin file directly?

**Comment by socialpreneur** [comment]
This (https://github.com/wp-media/wp-rocket/issues/3142#issuecomment-699957659) solution fixed a few sites, but not a complete solution unfortunately. But it's much better than WP Rocket's current code. Client sites are always different, and it's hard to figure out the exact cause of problem...

**Comment by roam92** [comment]
_> This fixed a few sites, but not a complete solution unfortunately._ 

Why do you say that? Can you give more details? Thanks!

**Comment by socialpreneur** [comment]
@roam92 It's hard to give the details as I have not figure this out the exact reason why it didn't work on certain sites. If I knew, I would have included some files that may be still conflicting with the fix. Due to NDA, I can't even put links to client sites here.

**Comment by roam92** [comment]
@socialpreneur – So to be clear, you made the change (locally) to the WP Rocket code proposed by @engahmeds3ed above, and what you found is that it fixed the mobile double-tap problem for some of the sites you tried it on, but not on all the sites?

For the sites you tested where the double-tap problem was resolved, was there any impact from the code change on the Delay JS functionality in WP Rocket, or did all that still work properly with the change?

From what I can tell, it seems to be an issue that selectively arises depending on the other JavaScript running on a site... I hope the WP Rocket team can take the time to isolate the circumstances when it arises and/or implement some sort of fix or workaround, because this can be a dealbreaker on iOS/iPadOS.

**Comment by paskalr1** [comment]
> After searching I got that there is an issue on IOS with Safari related to using `touchstart` or `touchmove` without using `touchend` which make the links double click issue ( you need to click on any point from the screen as first click then u can click on any link )
> 
> the following link tries to highlight the issue:
> https://stackoverflow.com/a/10340968
> 
> On customer site I made the following change and solved the issue:
> https://github.com/wp-media/wp-rocket/blob/7ec9d996f004707a57c4c8476090d5bd943051f0/assets/js/lazyload-scripts.js#L91-L99
> 
> 
> replace `touchmove` with `touchmove touchend`
> replace `touchstart` with `touchstart touchend`
> After checking with @hellofromtonya , she said:-
> 
> > This particular issue will need more investigation to find out why. Rather than adding a patch to the code for this specific edge case, why do they have the issue and why does this specific combination fix it?
> > I think the “why” here will be very valuable to ensure we don’t inject another problem for other customers.
> 
> so I will add needs: r&d label to be investigated.

I tried this and it works with my site

**Comment by roam92** [comment]
_> I tried this and it works with my site_

I tried it too, and it worked on my site as well...

It fixed the issue with double tap on iOS/iPadOS, and no Delay JS functionality was lost on any platform.

Not sure why this change hasn't already been made for everyone.

**Comment by alfonso100** [comment]
Another case: https://secure.helpscout.net/conversation/1499646121/260487?folderId=4130695

**Comment by JohnTorvik** [comment]
Related: https://secure.helpscout.net/conversation/1502403411/261233/

**Comment by DahmaniAdame** [comment]
Related ticket - https://secure.helpscout.net/conversation/1505662587/262130/

**Comment by socialpreneur** [comment]
> From what I can tell, it seems to be an issue that selectively arises depending on the other JavaScript running on a site...

I also think things there're some scripts that might work differently, even if WP Rocket adds this fix. It's just impossible to make this feature 100% working on any site.

**Comment by socialpreneur** [comment]
3.9 is released. I checked the updated file, but WP Rocket didn't include touchend fix for iOS. It would be great if someone form WP Rocket team explains why they skipped fixing this ongoing problem while some users reporting it fixes the issue.

**Comment by GeekPress** [comment]
@socialpreneur It hasn't been changed yet, it will be during the staggered process. Thanks for your comprehension and your patience.

**Comment by Mai-Saad** [comment]
Note for retest fix on browserstack: Couldn't reproduce using iPhone 12mini and iPhone 8 on Safari but was there on iPhone 11 and iPad Pro 12.9 on Safari (3.9)

**Comment by remyperona** [comment]
There will be an updated to this issue in 3.9.0.1, by adding additional touch events to the triggers for delay JS.

In our tests, it didn't prevent the double tap needed all the time, but there could be some improvements seen by our users.

We're keeping the issue open to keep track of the evolution.

**Comment by paskalr1** [comment]
> > After searching I got that there is an issue on IOS with Safari related to using `touchstart` or `touchmove` without using `touchend` which make the links double click issue ( you need to click on any point from the screen as first click then u can click on any link )
> > the following link tries to highlight the issue:
> > https://stackoverflow.com/a/10340968
> > On customer site I made the following change and solved the issue:
> > https://github.com/wp-media/wp-rocket/blob/7ec9d996f004707a57c4c8476090d5bd943051f0/assets/js/lazyload-scripts.js#L91-L99
> > 
> > replace `touchmove` with `touchmove touchend`
> > replace `touchstart` with `touchstart touchend`
> > After checking with @hellofromtonya , she said:-
> > > This particular issue will need more investigation to find out why. Rather than adding a patch to the code for this specific edge case, why do they have the issue and why does this specific combination fix it?
> > > I think the “why” here will be very valuable to ensure we don’t inject another problem for other customers.
> > 
> > 
> > so I will add needs: r&d label to be investigated.
> 
> I tried this and it works with my site

Google Analytics is not able to register mobile visitors if i implement the change. gtag/js in the delay scripts is causing the problem when the above mod is implemented. If i remove gtag/js from the list then the lighthouse score drops to the same level when the delay script is disabled. I have removed the mod until we find a new solution for this.

**Comment by pythoncreate** [comment]
Any updates here?  I tried to edit the aforementioned code, but mine seemingly looks different

**Comment by paskalr1** [comment]
installed the new version 3.9.0.2 and the pages were not rendering. I use wp bakery with the theme and has to exclude below scripts

jquery.min.js
js_composer_front.min.js

still double tap issue remains

website: https://www.celmetro.com/

**Comment by DahmaniAdame** [comment]
Issue still there on iPhone 12 Pro - iOS 14.5
Ticket - https://secure.helpscout.net/conversation/1543769478/271789/
@Mai-Saad would it be possible to test this one?

**Comment by Mai-Saad** [comment]
@DahmaniAdame There was an enhancement done on code but it isn't fixed 100% .. still can see it on some links on some sites. as mentioned here https://github.com/wp-media/wp-rocket/issues/3142#issuecomment-853129070

**Comment by GoldenGoldy** [comment]
I also had the double click/tap issue on IOS / Safari, on WP Rocket version 3.9.0.4
Managed to solve it with the info in this topic:D
I added the exclusion for twitter as suggested in a previous comment by @socialpreneur 
In my case I also added an exclusion for Facebook, as the issue was still present.
After adding the Facebook exclusion the issue is gone!
I also had the issue described here https://github.com/wp-media/wp-rocket/issues/3075 with Google Recaptcha so added an exclusion for that as well.
And I also removed 2 of the default exclusions:
`(?:wp-content|wp-includes)(.*)`
`/jquery-?[0-9.](.*)(.min|.slim|.slim.min)?.js`
, so far that works fine as well, no double click/tap issue.
Especially the first rule above excludes so many scripts that it's probably better to try and remove it and figure out which of the "internal" scripts should be excluded on an individual basis rather than using this catch-all rule.

My current exclusion list below for reference:
`js-(before|after)`
`google.com/recaptcha/api.js`
`platform.twitter.com/widgets.js`
`connect.facebook.net/en_US/sdk.js`

**Comment by roberthedlund** [comment]
> I also had the double click/tap issue on IOS / Safari, on WP Rocket version 3.9.0.4
> Managed to solve it with the info in this topic:D
> I added the exclusion for twitter as suggested in a previous comment by @socialpreneur
> In my case I also added an exclusion for Facebook, as the issue was still present.
> After adding the Facebook exclusion the issue is gone!
> I also had the issue described here #3075 with Google Recaptcha so added an exclusion for that as well.
> And I also removed 2 of the default exclusions:
> `(?:wp-content|wp-includes)(.*)`
> `/jquery-?[0-9.](.*)(.min|.slim|.slim.min)?.js`
> , so far that works fine as well, no double click/tap issue.
> Especially the first rule above excludes so many scripts that it's probably better to try and remove it and figure out which of the "internal" scripts should be excluded on an individual basis rather than using this catch-all rule.
> 
> My current exclusion list below for reference:
> `js-(before|after)`
> `google.com/recaptcha/api.js`
> `platform.twitter.com/widgets.js`
> `connect.facebook.net/en_US/sdk.js`

Thank you @GoldenGoldy this also worked for my site!!!! Awesome :)

**Comment by Alex43515** [comment]
Related ticket - https://secure.helpscout.net/conversation/1570565881/279394?folderId=4831552

**Comment by DahmaniAdame** [comment]
Potential fix using the [FastClick library](https://github.com/ftlabs/fastclick). 

Add the following to the header: 

```
<script src="https://cdnjs.cloudflare.com/ajax/libs/fastclick/1.0.6/fastclick.min.js" integrity="sha512-qH+R6YL4/40iiIrnN5aNZ1sEeEalNAdnzP9jfsxFPBdIslTkwUddkSazjVWhJ3f/3Y26QF6aql0xeneuVw0h/Q==" crossorigin="anonymous" referrerpolicy="no-referrer"></script>
<script>
if ('addEventListener' in document) {
	document.addEventListener('DOMContentLoaded', function() {
		FastClick.attach(document.body);
	}, false);
}
</script>
```

Exclude the following from Delay JavaScript Execution: 
```
fastclick
```

And test if the issue is fixed.

The fix worked on the following case: https://secure.helpscout.net/conversation/1567743453/278526

Requires further testing to confirm.

@Mai-Saad if you have a persistent double-tap on a testing environment, would you mind testing the above fix?

**Comment by mifrero** [comment]
Related ticket: https://secure.helpscout.net/conversation/1575244886/280804/
The fix provided by @DahmaniAdame worked on this case, but the website header disappears.

**Update:** The fix provided by @DahmaniAdame works fine, adding the script to the <body> that way solved the problem with the header, however this fix negatively impacts the load time, more info in the above ticket.

@Mai-Saad If you are checking this out, this case might be useful

**Comment by viobru** [comment]
Related ticket: https://secure.helpscout.net/conversation/1586259868/283864?folderId=3051575

**Comment by juricazuanovic** [comment]
Related ticket: https://secure.helpscout.net/conversation/1578504399/281872/

**Comment by camilamadronero-zz** [comment]
Related https://secure.helpscout.net/conversation/1613606249/290464/

**Comment by piotrbak** [comment]
Hello! We're still looking closely at this issue, of course. It's possible that in some cases the issue is not related to the Delay JS itself, but the [CSS code](https://stackoverflow.com/questions/32371431/hover-on-ios-mobile-devices-turns-into-double-touch-instead-of-hover):
```
iOS will not trigger a link click event on the first tap if the :hover state either:

Has a CSS transition animation
Reveals child content (such as a submenu, tooltip or ::before/::after element)
```

If you can confirm that the problem is not happening when Delay JS is enabled (or WP Rocket's optimizations bypassed), please post the URL of your site or send it to me piotr [at] wp-media.me.

**Comment by blogtutor** [comment]
Hi @piotrbak - I'm unclear if you're looking for a site that has the issue, or doesn't have the issue? But in case it's helpful, here's a client site that has it: https://cubesnjuliennes.com.

We currently have both Delay JS & Defer JS Enabled (with a few exclusions for her ConvertPro, her opt-in tool).  I'm able to reproduce the "must tap twice" issue on the homepage on my iPhone...both cached/optimized _and_ with `?nowprocket` appended to the URL.

Steps to reproduce:

1. Go to https://cubesnjuliennes.com or https://cubesnjuliennes.com?nowprocket on iPhone X Safari, iOS 14.7.1.
2. Scroll down to the "Latest Recipes" thumbnail images.
3. Tap on one. It will highlight and stay highlighted.
4. Tap on it again and it will load the link.

I'm able to then click the logo on her site to go back to the homepage and repeat the process. However, another one of my team members (also on iPhone X with iOS 14.7.1) only sees the issue on the first page load... 

Hope that helps?

**Comment by bemyhre** [comment]
I also *had* this issue until I turned off delay js at https://ramshacklepantry.com

It was very similar steps to reproduce that @blogtutor pointed out, but at my site AND ?nowprocket worked fine for me. 

Also one thing to note... I was getting this on individual recipe pages with the "jump to recipe" button, but only after I turned on my ad company's feature to insert an ad at the top of the recipe card. That feature hijacks both the button click and scroll (you can see it on my site right now).

When I turned it off, the Jump to Recipe button was fine. Without that feature on, the button is just an anchor tag element. With that feature on.. all sorts of js comes into play. 

I turned off delay js (which is not ideal) or ?nowprocket and it works no matter where.

**Comment by piotrbak** [comment]
Hello @blogtutor and @bemyhre 
Thank you for your feedbacks, they're very appreciated! At this moment there are a couple of possible culprits of this behaviour. Please find the troubleshooting steps below:
1. Double click is happening with and without Delay JS - it means that the problem is most likely related to the CSS and this [issue](https://stackoverflow.com/questions/32371431/hover-on-ios-mobile-devices-turns-into-double-touch-instead-of-hover).
2. If the problem is happening only with Delay JS enabled, then it's possible that one of the loaded scripts is involved. To check this, please exclude **all** inline and external scripts from Delay JS (to confirm this, please go to the source of the page and search for the `type="rocketlazyloadscript"`, there shouldn't be any occurrence of this string)
3. If the above won't help, we'd love to investigate the problem because it'll be pointing to our script itself. In this case, please send me message to my mailbox provided in the previous comment.
Thank you once again!

**Comment by camilamadronero** [comment]
Related https://secure.helpscout.net/conversation/1635656263/295727/

**Comment by mifrero** [comment]
Related ticket: https://secure.helpscout.net/conversation/1645076349/297810?folderId=2952229

**Comment by camilamadronero-zz** [comment]
Related https://secure.helpscout.net/conversation/1651016587/299237/

**Comment by alfonso100** [comment]
Related: https://secure.helpscout.net/conversation/1624889905/293173

**Comment by camilamadronero-zz** [comment]
New ticket https://secure.helpscout.net/conversation/1687816685/306024?folderId=5321805

**Comment by girlie** [comment]
New ticket: https://secure.helpscout.net/conversation/1692844093/307019/

**Comment by WordPresseur** [comment]
New ticket: https://secure.helpscout.net/conversation/1691278202/306676/

Update: the solution provide here https://github.com/wp-media/wp-rocket/issues/3142#issuecomment-882012665 worked.

**Comment by peixotorms** [comment]
Can't we just detect safari on load, check if the delay option is on and issue a click somewhere, so that the user will always have 1 click done right after page load?

**Comment by WordPresseur** [comment]
New ticket: https://secure.helpscout.net/conversation/1696313190/307777/

**Comment by roberthedlund** [comment]
I just wanted to let you guys know that, after updating to latest WP-rocket this issue is totally gone on two of my WooComerce sites. I didt see anything related to this in the change log? 😎👌

**Comment by webtrainingwheels** [comment]
https://secure.helpscout.net/conversation/1718385239/312899?folderId=377611
This case was solved with this; https://github.com/wp-media/wp-rocket/issues/3142#issuecomment-882012665

**Comment by DahmaniAdame** [comment]
Related to - https://secure.helpscout.net/conversation/1729817263/314706/
Using the FastClick fix worked.

**Comment by DahmaniAdame** [comment]
Related - https://secure.helpscout.net/conversation/1747586547/317449/
Using the FastClick fix worked.

**Comment by WordPresseur** [comment]
Related: https://secure.helpscout.net/conversation/1754094479/318724?folderId=1213662
This case was solved with this [comment](https://github.com/wp-media/wp-rocket/issues/3142#issuecomment-882012665)

**Comment by camilamadronero-zz** [comment]
New ticket: https://secure.helpscout.net/conversation/1765823739/320992/

**Comment by vmanthos** [comment]
**Related ticket:** https://secure.helpscout.net/conversation/1794828640/327600/

**Comment by TorbenLundsgaard** [comment]
This issue is clearly worse since version 3.10.9 where the delay script was updated

**Comment by vmanthos** [comment]
**Related ticket:** https://secure.helpscout.net/conversation/1828939828/334083?folderId=2135277

**Comment by NataliaDrause** [comment]
related: https://secure.helpscout.net/conversation/1821353471/332640?folderId=3864740

**Comment by DahmaniAdame** [comment]
Related - https://secure.helpscout.net/conversation/1820111756/332448/
The issue is more accentuated on `3.10.9` when compared with `3.10.8`. The click fails to trigger way more often vs occasionally on the previous version.

**Comment by webtrainingwheels** [comment]
Related with 3.10.9
https://secure.helpscout.net/conversation/1835702047/335384?folderId=377611

**Comment by viobru** [comment]
Related: https://secure.helpscout.net/conversation/1836254774/335539/

**Comment by juricazuanovic** [comment]
Related: https://secure.helpscout.net/conversation/1842091476/336583/

**Comment by worldwildweb** [comment]
related: https://secure.helpscout.net/conversation/1844169714/336844/

**Comment by viobru** [comment]
Related: https://secure.helpscout.net/conversation/1846834763/337294

**Comment by juricazuanovic** [comment]
Related: https://secure.helpscout.net/conversation/1848457954/337556

**Comment by girlie** [comment]
Related: https://secure.helpscout.net/conversation/1849069752/337633/

**Comment by juricazuanovic** [comment]
Related: https://secure.helpscout.net/conversation/1848498492/337563/

**Comment by girlie** [comment]
Related: https://secure.helpscout.net/conversation/1857224034/338785/

**Comment by tranngocminhhieu** [comment]
Since 3.10.9 the issue has came back, I use Jnews theme and has got the issue with mobile version.

**Comment by Garnet-Fox** [comment]
Since version 3.10.9, this is not only a bug of the Safari/Ios mobile version. There are also problems on Chrome Destop last version. The problem is the click before the delayed scripts are fully loaded. At the same time, if a click event occurs while loading delayed scripts, the click event may fire, or it may not fire after loading delayed scripts. No errors in the console. Rolling back the script to the previous version fixes the situation. Theme Wordpress Porto.

**Comment by piotrbak** [comment]
@tranngocminhhieu @Garnet-Fox Could you reach our support team and send the URLs? If you have a staging site it's even better. We'll investigate the issue further and provide the solution asap

**Comment by tranngocminhhieu** [comment]
> 
@piotrbak 

Here my demo site with new and old version

Version 3.11.1 -- Need twice click to open menu
https://huongdan.lucidgen.com/

Version 3.10.7 -- One click to open menu
https://lucidgen.com/

Hope team can fit this problem for Jnews theme in the next update!

**Comment by Garnet-Fox** [comment]
Yes, a good illustrative example https://huongdan.lucidgen.com /. Quickly click in the main menu, let's say links from Home to News and vice versa, several times. The link is not always followed, the click is lost.

**Comment by dazetoyo** [comment]
can confirm that version 3.10.7 does not have this problem. Wp-rocket team please stop messing up this script every couple of weeks... :)

**Comment by piotrbak** [comment]
@tranngocminhhieu This is what I'm experiencing with your example:
https://www.loom.com/share/63b708a94e6549e0a1e2fb72d2362d22

Scripts are loading around 1-4 seconds each, but the menu is loaded at the end. We're going to improve this, but the menu is loaded, could you confirm that?

@dazetoyo We're happy to investigate each issue for further improvements, however, without URL to your site where it's happening, it's not possible.

**Comment by tranngocminhhieu** [comment]
@piotrbak 
You need to test on an actual phone. I confirm that the menu can not open with one click on my iPhone. 😔

**Comment by juricazuanovic** [comment]
Related: https://secure.helpscout.net/conversation/1873508285/341405/

**Comment by dazetoyo** [comment]
@piotrbak 
there is nothing to investigate
compare version 3.10.7 with the version live right now and see what changes you made to the script that affects iphones.
And yes, tests have to be done with an actual iphone, either using browserstack or some emulator, chrome tools won't work.

**Comment by piotrbak** [comment]
**>there is nothing to investigate**
@dazetoyo thank you for your opinion.

**>compare version 3.10.7 with the version live right now and see what changes you made to the script that affects iphones.**
We're working on it and there is no one reason, but rather a couple of them. Sharing your website via our Support Team or here will be valueable to gather more data.

Could you also share what version of iOS are you using exactly?

**Comment by viobru** [comment]
Related: https://secure.helpscout.net/conversation/1888852368/343735/

**Comment by worldwildweb** [comment]
related: https://secure.helpscout.net/conversation/1891607252/344233

**Comment by DahmaniAdame** [comment]
Related - https://secure.helpscout.net/conversation/1903198858/345790/
Fastclick didn't help in this case.

**Comment by andyz02** [comment]
Do you already know when you will be able to fix the double click issue? We haven't been able to update from 3.10.8 on multiple sites yet due to the double click issue. Your support also received all our login credentials for checking already...

**Comment by tranngocminhhieu** [comment]
My solution is to switch to LiteSpeed Cache (my hosting is LiteSpeed Server); the performance is excellent. The problem was resolved.

I can not wait for the developer to fix the problem. So I don't have to spend money for WPRocket after switching.

**Comment by viobru** [comment]
Related: https://secure.helpscout.net/conversation/1921871383/348735/

**Comment by camilamadronero-zz** [comment]
Related https://secure.helpscout.net/conversation/1929478938/350276/

**Comment by GoldenGoldy** [comment]
Currently using version 3.11.3 (latest version) of WP Rocket and the issue still exists. When "Delay Javascript execution" is enabled, you need to double tap/click in Safari on IOS almost 100% of the time. Interestingly, even when adding exclusion rules that cover all of the .js scripts loaded on the page, the issue still occurs.  
When disabling "Delay Javascript execution" the issue is gone immediately, only one tap/click needed for any links.
Tested on iPhone 13 Pro with software version 15.5

Is there a solution coming in the near future or is WP Rocket ignoring this problem?

**Comment by roam92** [comment]
This ticket is almost two years old and still open...

**Comment by GoldenGoldy** [comment]
> This ticket is almost two years old and still open...

That's true!
Before the issue was manageable though, at least for me, I would figure out which scripts to add to the exclusions as only some (for example Facebook, Twitter, and Google reCaptcha) were giving issues. It would work correctly with a few scripts in the exclusions.
At the moment though, I can't use the delay Javascript function at all, as per my message above.

I see others mentioned that this more severe issue was introduced in 3.10.9 so I had some hope that WP Rocket would fix that particular issue that was introduced relatively recently, and which renders the delay Javascript functionality completely useless.

**Comment by piotrbak** [comment]
@GoldenGoldy Of course, we're working on this problem. There are a couple of different `kinds` of this issue, each of them is caused by different factors. Have you already sent us an URL or template where we could replicate it?

**Comment by andyz02** [comment]
We have the same issue with multiple websites. As there doesn't seem to be a fix from wp-rocket (we reported the issue already in March) we came up with a quick-fix in the mean time.

Basically we are simulation the first user interaction on iOS devices. This way we fix the double-click issue. Just put this code to the footer and add the script to the late execution exclusions:

<script id="fakeuserinteraction">
if(navigator.userAgent.match(/(iPhone|iPod|iPad)/i)) {
	window.onload = () => window.dispatchEvent(new Event('touchend'));
}
</script>

**Comment by GoldenGoldy** [comment]
Thanks for that workaround! I'll look into that. 
And let's hope WP Rocket will come up with a proper solution soon.

**Comment by roam92** [comment]
See also if the [workaround proposed above](https://github.com/wp-media/wp-rocket/issues/3142#issuecomment-699957659) in September 2020 by @engahmeds3ed works for you... It did for me.

**Comment by GoldenGoldy** [comment]
Great thanks will look into that as well! Not the greatest solution I guess to start editing the plugin's code but if nothing else helps it can be considered as temporary fix I suppose.

**Comment by viobru** [comment]
Related: https://secure.helpscout.net/conversation/1940106113/353887

**Comment by worldwildweb** [comment]
related: https://secure.helpscout.net/conversation/1943901877/354756/

**Comment by froddo7** [comment]
> We have the same issue with multiple websites. As there doesn't seem to be a fix from wp-rocket (we reported the issue already in March) we came up with a quick-fix in the mean time.
> 
> Basically we are simulation the first user interaction on iOS devices. This way we fix the double-click issue. Just put this code to the footer and add the script to the late execution exclusions:
> <script id="fakeuserinteraction"> if(navigator.userAgent.match(/(iPhone|iPod|iPad)/i)) { window.onload = () => window.dispatchEvent(new Event('touchend')); } </script>

Yes I can confirm this workaround works on our website with Tagdiv's Newspaper Pro Theme. No more double click on Iphones or Androids even. Thanks a lot!

**Comment by aharonyltd** [comment]
This solution is not good:

<script id = "fakeuserinteraction"> if (navigator.userAgent.match (/ (iPhone | iPod | iPad) / i)) {window.onload = () => window.dispatchEvent (new Event ('touchend')); } </script>

It causes all JS files to load

This means that the JavaScript Delay does not work with this code

**Comment by froddo7** [comment]
> This solution is not good:
> 
> 
> 
> <script id = "fakeuserinteraction"> if (navigator.userAgent.match (/ (iPhone | iPod | iPad) / i)) {window.onload = () => window.dispatchEvent (new Event ('touchend')); } </script>
> 
> 
> 
> It causes all JS files to load
> 
> 
> 
> This means that the JavaScript Delay does not work with this code

Can anyone else confirm this observation?

**Comment by aharonyltd** [comment]
> > This solution is not good:
> > <script id = "fakeuserinteraction"> if (navigator.userAgent.match (/ (iPhone | iPod | iPad) / i)) {window.onload = () => window.dispatchEvent (new Event ('touchend')); } </script>
> > It causes all JS files to load
> > This means that the JavaScript Delay does not work with this code
> 
> Can anyone else confirm this observation?

Check the site before you enter the code and after you enter:
https://www.browserstack.com/

Check the link in the GT
https://gtmetrix.com/reports/ganor.ussl.co.il/qIJdj94C/

Analysis options: iPhone XS Max/11 Pro/11 Pro Max/12 Pro Max

The code is currently working on the site and is disabling the JavaScript Delay

**Comment by viobru** [comment]
Related: https://secure.helpscout.net/conversation/1929331126/350239

**Comment by viobru** [comment]
Related: https://secure.helpscout.net/conversation/1952246494/356557/

**Comment by joejoe04** [comment]
Related: https://secure.helpscout.net/conversation/1952930497/356667/

**Comment by neilcthacker** [comment]
Also experiencing this issue. I have tried all of the fixes listed throughout this thread, but no luck.

It's a shame because the delay JS feature is by far the biggest impact on my site speed. Without it WP Rocket is fairly useless on my site.

**Comment by viobru** [comment]
Likely related: https://secure.helpscout.net/conversation/1953165220/356716

**Comment by joejoe04** [comment]
Related: https://secure.helpscout.net/conversation/1955395538/357227/

**Comment by juricazuanovic** [comment]
Related https://secure.helpscout.net/conversation/1955665009/357352

**Comment by DahmaniAdame** [comment]
Related - https://secure.helpscout.net/conversation/1953813548/356889

**Comment by froddo7** [comment]
Yes this Delay JS seems to be an IOS/Apple issue. Till this is fixed, we have moved to perfmatters plugin.

**Comment by neilcthacker** [comment]
> Yes this Delay JS seems to be an IOS/Apple issue. Till this is fixed, we have moved to perfmatters plugin.

So this issue is exclusive to WPRocket in your experience?

**Comment by froddo7** [comment]
> > Yes this Delay JS seems to be an IOS/Apple issue. Till this is fixed, we have moved to perfmatters plugin.
> 
> 
> 
> So this issue is exclusive to WPRocket in your experience?

No it's not exclusive to WP rocket but our theme works ok with perfmatters & delay js.

**Comment by viobru** [comment]
Related: https://secure.helpscout.net/conversation/1958425178/357855

**Comment by jivaneo** [comment]
Hi there,
So, not sure to understand, is this issue fixed?

**Comment by cafecafe** [comment]
Having the same issue on the client`s website: Storefront theme + Woocommerce + Elementor menu (same issue with default frontend.min.js too). Both iOS Chrome and Safari.

**Comment by mifrero** [comment]
Related: https://secure.helpscout.net/conversation/1971231894/360426/

**Comment by jivaneo** [comment]
> Related: https://secure.helpscout.net/conversation/1971231894/360426/

Thanks, but I don't have access to this conversation, can you post the solution here?

**Comment by DahmaniAdame** [comment]
Related - https://secure.helpscout.net/conversation/1966571981/359434/

**Comment by worldwildweb** [comment]
related: https://secure.helpscout.net/conversation/1982730694/362571

**Comment by DahmaniAdame** [comment]
Related - https://secure.helpscout.net/conversation/1981258356/362306/

**Comment by mifrero** [comment]
Related: https://secure.helpscout.net/conversation/1977835670/361715/

**Comment by froddo7** [comment]
Hi so has this issue been fixed by WP Rocket? I would appreciate some feedback from the company's developers and other users too. Thanks!

**Comment by camilamadronero-zz** [comment]
@froddo7 Sorry, this hasn't been resolved on WP Rocket's side yet. This issue will be closed when that's the case.
As of now, I can only suggest using the alternatives and workarounds available in the comments, it seems you've already tried one that worked. Other from that, once there is a fix integrated in the plugin, this issue will be closed. Thanks!

Related new case: https://secure.helpscout.net/conversation/1992349485/364651/.

**Comment by DahmaniAdame** [comment]
Related - https://secure.helpscout.net/conversation/1992728967/364749/

**Comment by alfonso100** [comment]
Related https://secure.helpscout.net/conversation/1990386574/364231/

**Comment by worldwildweb** [comment]
related: https://secure.helpscout.net/conversation/2000643761/366632/

**Comment by juricazuanovic** [comment]
Related: https://secure.helpscout.net/conversation/1999398636/366160

**Comment by juricazuanovic** [comment]
related: https://secure.helpscout.net/conversation/2016371168/370140/

**Comment by joejoe04** [comment]
Related: https://secure.helpscout.net/conversation/2015657841/369944/

**Comment by juricazuanovic** [comment]
Related: https://secure.helpscout.net/conversation/2030228006/373301

**Comment by joejoe04** [comment]
Possibly related: https://secure.helpscout.net/conversation/2030219993/373299/

**Comment by joejoe04** [comment]
Related: https://secure.helpscout.net/conversation/2030200182/373293/

**Comment by worldwildweb** [comment]
related: https://secure.helpscout.net/conversation/2012984187/369324/

**Comment by mifrero** [comment]
Related: https://secure.helpscout.net/conversation/2044517455/376256?folderId=2952229

**Comment by juricazuanovic** [comment]
Related: https://secure.helpscout.net/conversation/2049825199/377509/

**Comment by camilamadronero-zz** [comment]
Related https://secure.helpscout.net/conversation/2062261386/380391/

**Comment by worldwildweb** [comment]
related: https://secure.helpscout.net/conversation/2064913990/380967

**Comment by mifrero** [comment]
Related: https://secure.helpscout.net/conversation/2065747531/381145/

**Comment by joejoe04** [comment]
Related: https://secure.helpscout.net/conversation/2067133128/381469/

**Comment by DahmaniAdame** [comment]
Related - https://secure.helpscout.net/conversation/2057770178/379392/

**Comment by joejoe04** [comment]
Related: https://secure.helpscout.net/conversation/2074063973/383052/

**Comment by mikezielonka** [comment]
I hope one day this gets fixed.  Maybe its time to 🙏

**Comment by ollygaspar** [comment]
This has been an issue with WP-Rocket for over 2 years now and seriously compromises the monetization potential for content sites. Loading an extra outdated script (fastclick) is not a permanent solution. 

Pretty disappointing to see no traction regarding this after so many cases reported to support teams.

**Comment by GeekPress** [comment]
> This has been an issue with WP-Rocket for over 2 years now and seriously compromises the monetization potential for content sites. Loading an extra outdated script (fastclick) is not a permanent solution.
> 
> Pretty disappointing to see no traction regarding this after so many cases reported to support teams.

What a bad supposition you are doing here. It's not because we aren't replying or don't see changes here that means we aren't working on it. 

We do...

You don't see it as we have another private repository for our the delay js file.

**Comment by viobru** [comment]
Related: https://secure.helpscout.net/conversation/2082598568/385798

**Comment by alfonso100** [comment]
related https://secure.helpscout.net/conversation/2057252274/379274?folderId=2683093

**Comment by froddo7** [comment]
I'm not sure, but this worked for me (so I think, correct me if I'm wrong). Was posted on a Facebook WP Rocket Group.

I am guessing you can play around with the delayed milliseconds?

Do post feedback if this is a solution, as I am a bit clueless if this works, or not. Good luck!

Go to WP Rocket Settings, and Exclude timingDelayExecution from JS Delay list.

Place in functions.php (of theme) the following code
![wprocketcode](https://user-images.githubusercontent.com/27906279/206900823-4ffd0b44-4abb-4fce-b3aa-12e119ac5caa.jpg)

**Comment by peixotorms** [comment]
> I'm not sure, but this worked for me (so I think, correct me if I'm wrong). Was posted on a Facebook WP Rocket Group.
> 
> I am guessing you can play around with the delayed milliseconds?
> 
> Do post feedback if this is a solution, as I am a bit clueless if this works, or not. Good luck!
> 
> Go to WP Rocket Settings, and Exclude timingDelayExecution from JS Delay list.
> 
> Place in functions.php (of theme) the following code ![wprocketcode](https://user-images.githubusercontent.com/27906279/206900823-4ffd0b44-4abb-4fce-b3aa-12e119ac5caa.jpg)

This would obviously be the equivalent of disabling the delay js feature altogether.

**Comment by froddo7** [comment]
@peixotorms Actually that's where the timing delay number comes into play doesn't it? A previous version of this fake user interaction triggered without any delay. So that disabled the delay js feature. Here I am guessing you can delay this fake interaction? What do you think? Thanks!

**Comment by peixotorms** [comment]
You can check on pagespeed insights or gtmetrix.com for what scripts are loaded before and after this code.
The waiting time of 0.5 seconds will only delay the scripts for that time, hence they will all get loaded even without interaction, except that now the page is also 0.5 seconds slower that without js delay activated.

Also, it's possible for people to be rushing and be faster than 0.5 seconds, so it would not work.
The only real solution for the double click is to properly exclude the scripts that are required for the menu to open, from lazy loading.

**Comment by WordDom** [comment]
Hi, do you know how to put this code in the last version ?
/assets/js/lazyload-scripts.js doesn't exist anymore, only a minify file
 const instance = new RocketLazyLoadScripts( 
 	[ 
 		'keydown', 
 		'mouseover', 
 		'touchmove', 
 		'touchstart' 
 	], 
 	browser 
 );

**Comment by moisty70** [comment]
Why you don't you say near delay check that enabling will broke your site on iphones?

**Comment by wessiack** [comment]
this issue is now open for so long.
I don't understand why wp-rocket doesn't fix it.
We are still on version 3.10.8 (where this issue does not exist!) and can't update....
I do not understand why wp-rocket had a working version thill 3.10.8 and can't fix that one from there?

**Comment by piotrbak** [comment]
@wessiack One of the root causes was solved in the 3.12.5 version of WP Rocket. Have you already sent your examples to our support team, so we could investigate this further? It'd be appreciated.

**Comment by wessiack** [comment]
Thanks for your response @piotrbak we are on the latest version 3.12.4. on integration. Is 3.12.5 already officially released? I don't see these in the plugin-update list...

**Comment by moisty70** [comment]
Same here, only 3.12.4 available. Hope that 3.12.5 solves this issue.

**Comment by piotrbak** [comment]
@wessiack @moisty70 3.12.5 is in the staggered release, you should see the update notification in upcoming days.

```
Hope that 3.12.5 solves this issue, but I'm not confident as is something that happens with many websites not only with wprocket.
```
We're fixing scenarios that are possible to be improved on our end. Here we're dependent on iOS browser implementations. For any regression cases (it used to work but it's not working in the newer version), the best solution is to contact our support team, we'll investigate each website separately finding the root cause.

**Comment by jojoba007** [comment]
Same issue here, WP 6.2, OxygenBuilder 4.5, MultiSite, Wp Rocket 3.12.6.1

**Comment by blw** [comment]
I don't see this fixed, I am on version 3.13.0.1

**Comment by mikezielonka** [comment]
Same! Still same issue as it has been for the past 3 years. Turning the JS delay has been the only solution.

**Comment by blw** [comment]
Any update on when this will be fixed? This is really a great feature but it's sad we can't enable it as most users are on mobile safari.

**Comment by jojoba007** [comment]
Any update on the matter? Issue still persists

**Comment by uzelacgs** [comment]
I have same issue with Newspaper theme!

**Comment by sentenza13** [comment]
Same issue, it will be really great if this can be fixed

**Comment by wessiack** [comment]
same issue (Flatsome Theme, WPML and  WP-Rocket Version 3.14.4.2) - committed by WP-Rocket 2nd level tech support!
Hope the WP-Rocket dev team will find a fix for that soon.

**Comment by lea-pw** [comment]
+1

**Comment by DOgunrinde3** [comment]
+1

**Comment by anandwebenix** [comment]
Hello, We are also facing this issue with our clients website. Please do the needful asap as this is critical.

**Comment by alexandrecorroy** [comment]
Hello, Any news about this problem ?

**Comment by zkingdesign** [comment]
This is freaking irresponsable, 3 years in and the issue still exists. Why can't it just be fixed? It's not like we're using a free plugin, we deserve better support, especially when sites break! If you can't fix it at least add a warning to the admin dashboard that states if you enable the javascript delay functionality the website may become unusable on safari.

**Comment by ProdanCristian** [comment]
Greetings everyone! For those who are still seeking a solution, I've created a tutorial that might help. It's been effective for my website, which boasts a 99 score on PageSpeed, and I've encountered no issues with delayed JavaScripts.
 
This is the solution: https://prodanadvisor.com/double-click-issue-on-ios-safari-with-wp-rocket/

**Comment by peixotorms** [comment]
I had the same issue, so I moved on from wp rocket long time ago... but also to be fair, people are forgetting, that when you delay all javascript your are effectively turning it off and it's not downloaded, until there is interaction. 

If the site was developed with a lot of plugins and is heavy with a lot of javascript, then obviously if you directly go and click the menu (how many people do that?) it first needs to download the scripts, and only then it will try to open the menu. Normal behaviour is to move the mouse, or tap somewhere else before going to the menu... so I think the issue shouldn't happen for those people, but if you are testing and directly clicking the menu, then the bug is going to happen.

Some scripts from different themes and plugins simply do not allow it to work due to how they are coded, or they have not being excluded because wp rocket doesn't recognize them.... I saw this coming when they switched from manually turning on the delay javascript option, to always on by default... so poor oversight.

They have to now exclude every possible script from all themes and plugins that have a menu, which also means, a lot of work and lower performance in the tests.

**Comment by peixotorms** [comment]
> Greetings everyone! For those who are still seeking a solution, I've created a tutorial that might help. It's been effective for my website, which boasts a 99 score on PageSpeed, and I've encountered no issues with delayed JavaScripts.
> 
> This is the solution: https://prodanadvisor.com/double-click-issue-on-ios-safari-with-wp-rocket/

That is never going to be a solution... you are basically excluding Safari from the optimization, so while you score well on pagespeed (because it's not safari), all your safari users will not be optimized. 

That also means, regardless of having 100 on pagespeed, you are still going to fail the web vitals metrics if you have a significant amount of users using safari. Web vitals are not from pagespeed, they are the average speeds experienced by all users.

**Comment by ollygaspar** [comment]
While I've stopped using WP-Rocket, I was still experiencing the double click issue on IOS. 

I created a script that resolved this issue on my website. Here:

;(function(document, window) {
    'use strict';

    function isIOS() {
        return /iP(ad|hone|od)/.test(navigator.userAgent) && !window.MSStream;
    }

    if (!isIOS()) {
        return;
    }

    var touchStartX = 0, touchStartY = 0, touchStartTime = 0;
    var firstTouch = true; // Flag to track the first touch

    function onTouchStart(event) {
        if (!firstTouch) return; // Ignore if not the first touch

        var touch = event.touches[0];
        touchStartX = touch.pageX;
        touchStartY = touch.pageY;
        touchStartTime = event.timeStamp;
    }

    function onTouchEnd(event) {
        if (!firstTouch) return; // Ignore if not the first touch

        var touch = event.changedTouches[0];
        var moveX = Math.abs(touch.pageX - touchStartX);
        var moveY = Math.abs(touch.pageY - touchStartY);
        var duration = event.timeStamp - touchStartTime;

        if (moveX < 10 && moveY < 10 && duration < 200) {
            var targetElement = event.target;

            // Apply to links (including those within SVGs) and buttons
            if (targetElement.tagName.toLowerCase() === 'a' || 
                targetElement.tagName.toLowerCase() === 'button' ||
                targetElement.closest('a') ||
                targetElement.closest('button')) {

                var clickEvent = new MouseEvent('click', { bubbles: true, cancelable: true });
                targetElement.dispatchEvent(clickEvent);
                event.preventDefault(); // Prevent the default touchend behavior
            }

            firstTouch = false; // Reset the flag after the first touch
        }
    }

    document.addEventListener('touchstart', onTouchStart, false);
    document.addEventListener('touchend', onTouchEnd, false);

})(document, window);

**Comment by appelgran** [comment]
I had this double-tap issue caused by lightSlider due to its use of touchstart and touchmove that somehow triggered the bug in iOS. After thorough investigation I could resolve my issue with this piece of script (added to the beginning of body):

    <script type="module">document.addEventListener("touchstart", function(){});</script>

It might as well work for your wp-rocket issues too.

**Comment by Zanyprogressive** [comment]
@appelgran It fixed the issue for me! Thank you!

**Comment by betayazilim** [comment]
I had same issue and go back to perfmatters, they have a solution for this.

**Comment by jekayode** [comment]
**Related ticket:** https://secure.helpscout.net/conversation/2759456594/522480?folderId=8127831

**Comment by markonikolic985** [comment]
**Related ticket:** https://secure.helpscout.net/conversation/2828746699/536957?viewId=424682

**Comment by piotrbak** [comment]
The issue should be solved in 3.18.1 version of WP Rocket. The release is part of staggered process, it'll be soon available for everyone
